File: chains/cpp/fhevm/src/service.cpp
// Copyright (C) 2026, Lux Industries Inc. All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause-Eco

#include "lux/fhevm/service.hpp"

namespace lux::fhevm {
namespace {

CiphertextView to_view(const CiphertextRecord& r) {
    CiphertextView v;
    v.handle = hex(view(r.handle));
    v.owner = account_string(r.owner);
    v.scheme = r.scheme;
    v.digest = hex(view(r.digest));
    v.type = r.type;
    v.level = r.level;
    v.epoch = r.epoch;
    v.size = r.size;
    v.registered_at = r.registered_at;
    v.chain_id = id_string(r.chain_id);
    return v;
}

}  // namespace

Result<Account> account_from_hex(std::string_view s) {
    Bytes b;
    if (!from_hex(s, &b)) return fail(Err::InvalidPayload, "address is not hex");
    Account a{};
    if (b.size() != a.size()) return fail(Err::InvalidPayload, "address is the wrong width");
    std::copy(b.begin(), b.end(), a.begin());
    return a;
}

Result<Id> hash32(std::string_view s) {
    Bytes b;
    if (!from_hex(s, &b)) return fail(Err::InvalidPayload, "identifier is not hex");
    Id id{};
    if (b.size() != id.size()) return fail(Err::InvalidPayload, "identifier is the wrong width");
    std::copy(b.begin(), b.end(), id.begin());
    return id;
}

std::string_view Service::operation_name(std::uint8_t type) {
    switch (type) {
        case kTxRegisterCiphertext: return "registerCiphertext";
        case kTxGrantPermit: return "grantPermit";
        case kTxRevokePermit: return "revokePermit";
        case kTxRequestDecrypt: return "requestDecrypt";
        case kTxFulfillDecrypt: return "fulfillDecrypt";
        case kTxAdvanceEpoch: return "advanceEpoch";
        default: return "";
    }
}

Result<Id> Service::submit_transaction(ByteView raw) const {
    auto tx = parse_transaction(raw);
    if (!tx) return std::unexpected(tx.error());
    return vm_->submit_tx(*tx);
}

Result<CiphertextView> Service::ciphertext(std::string_view handle_hex) const {
    auto h = hash32(handle_hex);
    if (!h) return std::unexpected(h.error());
    const CiphertextRecord* rec = vm_->ciphertext(*h);
    if (rec == nullptr) return fail(Err::CiphertextNotFound);
    return to_view(*rec);
}

Result<std::vector<CiphertextView>> Service::ciphertexts(std::string_view owner_hex,
                                                         std::string_view scheme) const {
    std::vector<CiphertextView> out;
    Account owner{};
    bool by_owner = false;
    if (!owner_hex.empty()) {
        auto a = account_from_hex(owner_hex);
        if (!a) return std::unexpected(a.error());
        owner = *a;
        by_owner = true;
    }
    for (const CiphertextRecord* rec : vm_->ciphertexts()) {
        if (by_owner && rec->owner != owner) continue;
        if (!scheme.empty() && rec->scheme != scheme) continue;
        out.push_back(to_view(*rec));
    }
    return out;
}

Result<PermitView> Service::permit(std::string_view permit_id_hex) const {
    auto id = hash32(permit_id_hex);
    if (!id) return std::unexpected(id.error());
    const PermitRecord* rec = vm_->permit(*id);
    if (rec == nullptr) return fail(Err::PermitNotFound);
    PermitView v;
    v.permit_id = hex(view(rec->permit_id));
    v.handle = hex(view(rec->handle));
    v.grantor = account_string(rec->grantor);
    v.grantee = account_string(rec->grantee);
    v.operations = rec->operations;
    v.expiry = rec->expiry;
    v.status = rec->status;
    v.created_at = rec->created_at;
    v.chain_id = id_string(rec->chain_id);
    return v;
}

Result<DecryptView> Service::decrypt(std::string_view request_id_hex) const {
    auto id = hash32(request_id_hex);
    if (!id) return std::unexpected(id.error());
    const DecryptRecord* rec = vm_->decrypt(*id);
    if (rec == nullptr) return fail(Err::RequestNotFound);

    // A pending request past its expiry can no longer be answered — check_auth
    // refuses every attestation to it — so it is reported expired. The STORED
    // status stays pending because nothing ever answered it; expiry is a fact
    // about the clock, and this is the one place that reads the clock to say so.
    RequestStatus status = rec->status;
    if (status == RequestStatus::Pending && rec->expiry != 0 &&
        vm_->clock().now() > rec->expiry) {
        status = RequestStatus::Expired;
    }

    DecryptView v;
    v.request_id = hex(view(rec->request_id));
    v.handle = hex(view(rec->ciphertext_handle));
    v.requester = account_string(rec->requester);
    v.permit_id = hex(view(rec->permit_id));
    v.callback = hex(view(rec->callback));
    v.selector = hex(view(rec->callback_selector));
    v.epoch = rec->epoch;
    v.status = std::string(request_status_string(status));
    v.expiry = rec->expiry;
    v.created_at = rec->created_at;
    v.completed_at = rec->completed_at;
    v.source_chain = id_string(rec->source_chain);
    if (rec->status == RequestStatus::Completed) {
        v.result_handle = hex(view(rec->result_handle));
    }
    if (const EpochRecord* ep = vm_->epoch(rec->epoch)) v.threshold = ep->threshold;
    for (const auto& a : rec->attestations) {
        v.attestations.push_back(AttestationView{account_string(a.member), hex(view(a.value))});
    }
    return v;
}

Result<CommitteeView> Service::committee(std::uint64_t epoch) const {
    const EpochRecord* rec = vm_->epoch(epoch);
    if (rec == nullptr) return fail(Err::EpochNotFound);
    CommitteeView v;
    v.epoch = rec->epoch;
    v.threshold = rec->threshold;
    for (const auto& m : rec->committee) {
        v.members.push_back(CommitteeMemberView{node_id_string(m.node_id),
                                                hex(view(m.public_key)), m.weight, m.index});
    }
    return v;
}

CommitteeView Service::current_committee() const {
    auto v = committee(vm_->current_epoch_number());
    if (!v) return CommitteeView{vm_->current_epoch_number(), 0, {}};
    return *v;
}

Result<Service::Balance> Service::balance(std::string_view address_hex) const {
    auto acct = account_from_hex(address_hex);
    if (!acct) return std::unexpected(acct.error());
    auto bal = vm_->balance(*acct);
    if (!bal) return std::unexpected(bal.error());
    auto burned = vm_->burned();
    if (!burned) return std::unexpected(burned.error());
    return Balance{*bal, *burned};
}

Result<HealthReport> Service::health() const { return vm_->health(); }

Service::PublicParams Service::public_params() const {
    // lattigo's ExampleParameters128BitLogN14LogQP438 at the runtime's default
    // threshold configuration: ring dimension 2^14, log(QP) 435, default scale
    // 2^45. Pinned here, and pinned again in the test, so a change to the
    // runtime's parameters shows up as a failure rather than as two networks
    // encrypting under different moduli.
    PublicParams p;
    p.epoch = vm_->current_epoch_number();
    p.log_n = 14;
    p.log_qp = 435;
    p.log_scale = 45;
    p.chain_id = id_string(vm_->chain_id());
    if (const EpochRecord* ep = vm_->epoch(p.epoch)) {
        p.threshold = ep->threshold;
        p.public_key = hex(view(ep->public_key));
    }
    return p;
}

FeeSchedule Service::fee_schedule() const {
    FeeSchedule out;
    out.gas_price = kGasPrice;
    for (std::uint8_t op : {kTxRegisterCiphertext, kTxGrantPermit, kTxRevokePermit,
                            kTxRequestDecrypt, kTxFulfillDecrypt, kTxAdvanceEpoch}) {
        if (uses_scheme(op)) {
            for (std::string_view scheme : schemes()) {
                Transaction tx;
                tx.type = op;
                tx.scheme = std::string(scheme);
                auto g = gas_for(tx);
                auto f = fee_for(tx);
                out.entries.push_back(FeeScheduleEntry{std::string(operation_name(op)),
                                                       std::string(scheme), g ? *g : 0,
                                                       f ? *f : 0});
            }
            continue;
        }
        Transaction tx;
        tx.type = op;
        auto g = gas_for(tx);
        auto f = fee_for(tx);
        out.entries.push_back(
            FeeScheduleEntry{std::string(operation_name(op)), "", g ? *g : 0, f ? *f : 0});
    }
    return out;
}

}  // namespace lux::fhevm

File: chains/rust/fhevm/src/service.rs
// SPDX-License-Identifier: BSD-3-Clause-Eco
// Copyright (C) 2019-2026, Lux Industries Inc. All rights reserved.

//! The read surface, and the one way to change anything.
//!
//! Mutating operations are submitted as CLIENT-SIGNED transactions and take
//! effect only through fee-settled consensus blocks. Everything else here is a
//! read-only query of PUBLIC state. There is no endpoint that decrypts, none that
//! returns a ciphertext body, and no "fee" integer in any request — a fee is never
//! an unbacked number a caller writes down, it is gas metered and burned from the
//! payer's on-chain balance inside consensus.
//!
//! Two of the methods report the FHE RUNTIME's configuration rather than F's
//! ledger — the public parameters and the seated committee — and answer in the
//! runtime's vocabulary, so a client that understands the runtime understands F.
//! The rest describe F's ledger and answer in F's own views. That is the whole
//! rule: runtime types for runtime facts, F's views for F's records.

use serde_json::{json, Value};

use crate::block::Block;
use crate::error::{Error, Result};
use crate::fhe;
use crate::gas;
use crate::ids;
use crate::state::{CiphertextRecord, DecryptRecord, PermitRecord};
use crate::vm::{account_from_hex, Vm};
use crate::wire;

/// The PUBLIC view of a registered encrypted value. It carries the digest of the
/// ciphertext body so a client can check a body it fetched from off-chain
/// storage; it does not carry the body, because F never has it.
pub fn ciphertext_view(r: &CiphertextRecord) -> Value {
    json!({
        "handle": hex::encode(r.meta.handle),
        "owner": ids::short_id_string(&r.meta.owner),
        "scheme": r.scheme,
        "digest": hex::encode(r.digest),
        "type": r.meta.kind,
        "level": r.meta.level,
        "epoch": r.meta.epoch,
        "size": r.meta.size,
        "registeredAt": r.meta.registered_at,
        "chainId": ids::id_string(&r.meta.chain_id),
    })
}

/// The PUBLIC view of a capability grant.
pub fn permit_view(r: &PermitRecord) -> Value {
    json!({
        "permitId": hex::encode(r.permit.permit_id),
        "handle": hex::encode(r.permit.handle),
        "grantor": ids::short_id_string(&r.permit.grantor),
        "grantee": ids::short_id_string(&r.permit.grantee),
        "operations": r.permit.operations,
        "expiry": r.permit.expiry,
        "status": r.status,
        "createdAt": r.permit.created_at,
        "chainId": ids::id_string(&r.permit.chain_id),
    })
}

/// The PUBLIC view of a threshold-decryption request. It has no plaintext field:
/// the plaintext goes to the requester's callback off-chain, and F records only
/// the handle the committee agreed the decryption produced.
pub fn decrypt_view(vm: &Vm, r: &DecryptRecord) -> Value {
    // A pending request past its expiry can no longer be answered — every
    // attestation to it is refused — so it is reported expired. The stored status
    // stays pending because nothing ever answered it; expiry is a fact about the
    // clock, and this is the one place that reads the clock to say so.
    let mut status = r.request.status;
    if status == fhe::RequestStatus::Pending
        && r.request.expiry != 0
        && vm.clock.time().unix() > r.request.expiry
    {
        status = fhe::RequestStatus::Expired;
    }
    let attestations: Vec<Value> = r
        .attestations()
        .iter()
        .map(|a| {
            json!({
                "member": ids::short_id_string(&a.member),
                "value": hex::encode(a.value),
            })
        })
        .collect();
    let threshold = vm.epoch(r.request.epoch).map(|e| e.info.threshold).unwrap_or(0);
    let mut view = json!({
        "requestId": hex::encode(r.request.request_id),
        "handle": hex::encode(r.request.ciphertext_handle),
        "requester": ids::short_id_string(&r.request.requester),
        "permitId": hex::encode(r.permit_id),
        "callback": hex::encode(r.request.callback),
        "selector": hex::encode(r.request.callback_selector),
        "epoch": r.request.epoch,
        "status": status.as_str(),
        "attestations": attestations,
        "threshold": threshold,
        "expiry": r.request.expiry,
        "createdAt": r.request.created_at,
        "sourceChain": ids::id_string(&r.request.source_chain),
    });
    if r.request.status == fhe::RequestStatus::Completed {
        view["resultHandle"] = json!(hex::encode(r.request.result_handle));
        view["completedAt"] = json!(r.request.completed_at);
    }
    view
}

/// Decodes a hex-encoded 32-byte identifier — a handle, a permit id, a request id.
/// One decoder for all three, so none of them can drift into accepting a length
/// the others reject.
pub fn hash32(s: &str) -> Result<[u8; 32]> {
    let body = s.strip_prefix("0x").unwrap_or(s);
    let b = hex::decode(body).map_err(|e| Error::InvalidPayload(e.to_string()))?;
    if b.len() != 32 {
        return Err(Error::InvalidPayload(format!("identifier is {} bytes, not 32", b.len())));
    }
    let mut out = [0u8; 32];
    out.copy_from_slice(&b);
    Ok(out)
}

fn param<'a>(params: &'a Value, name: &str) -> Option<&'a Value> {
    params.get(name)
}

fn string_param(params: &Value, name: &str) -> String {
    param(params, name).and_then(|v| v.as_str()).unwrap_or_default().to_string()
}

/// Answers one call. The seam speaks JSON-RPC, so this is where a caller's request
/// becomes a query and a record becomes a view.
pub fn call(vm: &Vm, method: &str, params: &Value) -> Result<Value> {
    match method {
        // ---- the one mutating call: submit a client-signed transaction ----
        "fchain.submitTransaction" => {
            let raw_hex = string_param(params, "tx");
            let body = raw_hex.strip_prefix("0x").unwrap_or(&raw_hex);
            let raw = hex::decode(body).map_err(|e| Error::InvalidPayload(e.to_string()))?;
            let tx = wire::parse_transaction(&raw)?;
            let id = vm.submit_tx(&tx)?;
            Ok(json!({ "txId": ids::id_string(&id) }))
        }

        // ---- FHE runtime facts ----
        "fchain.getPublicParams" => {
            let epoch = vm.current_epoch();
            let p = fhe::DEFAULT_THRESHOLD_PARAMS;
            let mut out = json!({
                "epoch": epoch,
                "logN": p.log_n,
                "logQP": p.log_qp,
                "logScale": p.log_scale,
                "chainId": ids::id_string(&vm.chain_id),
                "threshold": 0,
                "publicKey": "",
            });
            if let Some(rec) = vm.epoch(epoch) {
                out["threshold"] = json!(rec.info.threshold);
                out["publicKey"] = json!(hex::encode(&rec.info.public_key));
            }
            Ok(out)
        }

        "fchain.getCommittee" => {
            let n = match param(params, "epoch").and_then(|v| v.as_u64()) {
                Some(n) => n,
                None => vm.current_epoch(),
            };
            let rec = vm.epoch(n).ok_or(Error::EpochNotFound)?;
            let members: Vec<Value> = rec
                .committee()
                .iter()
                .map(|m| {
                    json!({
                        "nodeId": ids::node_id_string(&m.node_id),
                        "publicKey": hex::encode(&m.public_key),
                        "weight": m.weight,
                        "index": m.index,
                    })
                })
                .collect();
            Ok(json!({
                "epoch": rec.info.epoch,
                "threshold": rec.info.threshold,
                "members": members,
            }))
        }

        // ---- F's ledger ----
        "fchain.getCiphertext" => {
            let h = hash32(&string_param(params, "handle"))?;
            let rec = vm.ciphertext(&h).ok_or(Error::CiphertextNotFound)?;
            Ok(json!({ "ciphertext": ciphertext_view(&rec) }))
        }

        "fchain.listCiphertexts" => {
            let owner_hex = string_param(params, "owner");
            let scheme = string_param(params, "scheme");
            // A filter whose address does not decode is an error, not an empty
            // listing: an empty listing reads as "this owner has nothing".
            let owner = if owner_hex.is_empty() {
                None
            } else {
                Some(
                    account_from_hex(&owner_hex)
                        .map_err(Error::InvalidPayload)?,
                )
            };
            let mut out: Vec<Value> = Vec::new();
            let mut records = vm.ciphertexts();
            // A listing is a read of a map, and a map has no order. Sorting by
            // handle gives one, so two nodes answering the same query answer the
            // same way.
            records.sort_by_key(|r| r.meta.handle);
            for rec in records {
                if let Some(o) = owner {
                    if rec.meta.owner != o {
                        continue;
                    }
                }
                if !scheme.is_empty() && rec.scheme != scheme {
                    continue;
                }
                out.push(ciphertext_view(&rec));
            }
            Ok(json!({ "total": out.len(), "ciphertexts": out }))
        }

        "fchain.getPermit" => {
            let id = hash32(&string_param(params, "permitId"))?;
            let rec = vm.permit(&id).ok_or(Error::PermitNotFound)?;
            Ok(json!({ "permit": permit_view(&rec) }))
        }

        "fchain.getDecrypt" => {
            let id = hash32(&string_param(params, "requestId"))?;
            let rec = vm.decrypt(&id).ok_or(Error::RequestNotFound)?;
            Ok(json!({ "request": decrypt_view(vm, &rec) }))
        }

        // ---- accounts ----
        "fchain.balance" => {
            let acct = account_from_hex(&string_param(params, "address"))
                .map_err(Error::InvalidPayload)?;
            // The two numbers are read independently, and a failure on either is
            // reported. Burned supply coming back zero on a failed read is
            // indistinguishable from a chain that has never settled a fee.
            let bal = vm.balance(&acct)?;
            let burned = vm.burned()?;
            Ok(json!({ "balanceNLux": bal, "burnedNLux": burned }))
        }

        // ---- diagnostics ----
        "fchain.health" => {
            let (healthy, details) = vm.health_check()?;
            let mut map = serde_json::Map::new();
            for (k, v) in details {
                map.insert(k, json!(v));
            }
            Ok(json!({ "healthy": healthy, "details": Value::Object(map) }))
        }

        "fchain.feeSchedule" => {
            let mut entries: Vec<Value> = Vec::new();
            for (op, name) in gas::OP_NAMES {
                if gas::uses_scheme(op) {
                    for (scheme, _) in gas::SCHEME_GAS {
                        let tx = crate::transaction::Transaction {
                            tx_type: op,
                            scheme: scheme.as_bytes().to_vec(),
                            ..crate::transaction::Transaction::default()
                        };
                        let g = gas::gas_for(&tx)?;
                        entries.push(json!({
                            "operation": name,
                            "scheme": scheme,
                            "gas": g,
                            "feeNLux": crate::fee::cost(g, gas::GAS_PRICE)?,
                        }));
                    }
                    continue;
                }
                let tx = crate::transaction::Transaction {
                    tx_type: op,
                    ..crate::transaction::Transaction::default()
                };
                let g = gas::gas_for(&tx)?;
                entries.push(json!({
                    "operation": name,
                    "scheme": "",
                    "gas": g,
                    "feeNLux": crate::fee::cost(g, gas::GAS_PRICE)?,
                }));
            }
            Ok(json!({ "gasPriceNLuxPerGas": gas::GAS_PRICE, "entries": entries }))
        }

        other => Err(Error::InvalidPayload(format!("the method {other} does not exist"))),
    }
}

/// The block the seam hands back, described the way the seam describes one.
pub fn block_summary(b: &Block) -> Value {
    json!({
        "id": ids::id_string(&b.id),
        "parentId": ids::id_string(&b.parent_id),
        "height": b.height,
        "timestamp": b.timestamp.unix(),
        "txs": b.transactions.len(),
    })
}
